### PR TITLE
Initial commit of new intersection

### PR DESCRIPTION
**Pull request by hbs**

### CODE REVIEW

File: src/main/java/com/geoxp/geo/GeoXPShapeHelper.java [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+import java.util.Arrays;
+import java.util.BitSet;
+
+import com.geoxp.GeoXPLib;
+import com.geoxp.GeoXPLib.GeoXPShape;
+
+public class GeoXPShapeHelper {
+  
+  private static long[] RESOLUTION_MASKS;
```
Comment (ftence) [inline src/main/java/com/geoxp/geo/GeoXPShapeHelper.java]: Should be `final`.

File: src/main/java/com/geoxp/geo/GeoXPShapeHelper.java [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    return null != intersection(a, b, true);
+  }
+  
+  public static GeoXPShape intersection(GeoXPShape a, GeoXPShape b) {
+    return intersection(a, b, false);
+  }
+  
+  private static GeoXPShape intersection(GeoXPShape a, GeoXPShape b, boolean check) {
+       
+   long[] acells = GeoXPLib.getCells(a);
```
Comment (ftence) [inline src/main/java/com/geoxp/geo/GeoXPShapeHelper.java]: Indent problem: from here there is a missing space at the beginning of each line.

File: src/main/java/com/geoxp/geo/GeoXPShapeHelper.java [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+         }
+         long cell = acells[i] & RESOLUTION_MASKS[res];
+         // Iterate over the resolutions finer or equal to res in the other shape
+         for (int subres = res; subres < 16; subres++) {
+           if (-1 == bstart[subres]) {
+             continue;
+           }
+           // attempt to find the insertion point for the subcell 0 at subres
+           long subcell = cell | (((long) subres) << 60);
+           int idx = Arrays.binarySearch(bcells, bstart[subres], bend[subres] + 1, subcell);
```
Comment (ftence) [inline src/main/java/com/geoxp/geo/GeoXPShapeHelper.java]: This signature of `Arrays.binarySearch` would require you to bump from Java 5 to Java 6.
As a result, you should update build.gradle with, for instance,
```
sourceCompatibility = 1.8
targetCompatibility = 1.8
```

File: src/main/java/com/geoxp/geo/GeoXPShapeHelper.java [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+
+   for (int i = 1; i < 16; i++) {
+     bend[i] = -1;
+     if (-1 != bstart[i]) {
+       // Find the start of the next index
+       for (int j = i + 1; j < 16; j++) {
+         if (-1 != bstart[j]) {
+           bend[i] = bstart[j] - 1;
+           break;
+         }
```
Comment (ftence) [inline src/main/java/com/geoxp/geo/GeoXPShapeHelper.java]: Ditto.

File: src/main/java/com/geoxp/geo/GeoXPShapeHelper.java [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+   
+   for (int i = 1; i < 16; i++) {
+     aend[i] = -1;
+     if (-1 != astart[i]) {
+       // Find the start of the next index
+       for (int j = i + 1; j < 16; j++) {
+         if (-1 != astart[j]) {
+           aend[i] = astart[j] - 1;
+           break;
+         }
```
Comment (ftence) [inline src/main/java/com/geoxp/geo/GeoXPShapeHelper.java]: This won't work because `aend[i] = astart[j] - 1` does not hold for `i < 8` and `j >= 8`.

**Review by ftence — CHANGES_REQUESTED** [review]

**Comment by hbs** [comment]
PTAL

**Comment by hbs** [comment]
PTAL

**Comment by ftence** [comment]
Here is a test on the intersection of blue and red shapes giving the orange result, with individual cells only displayed for the orange shape. Everything seems to work. Permutation test is also OK.
image

**Review by ftence — APPROVED** [review]

**Pull request merged**